dvc/repo/__init__.py

```diff
@@ -583,3 +583,20 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_val, exc_tb):
         self._reset()
         self.scm.close()
+
+
+class MemoryRepo(Repo):
```

Comment:
I don't think we need this. We just need to allow using memory url in the config. E.g.

```
[cache]
    local = memodb
[remote "memodb"]
    url = "memory://..."
```

Reply:
This is a pretty small piece of code that is not used right now anywhere and is rather hacky. Let's introduce it with the testing changes in the future.

Reply:
The memory url is just a proper way to support this. I'm not saying it is going to be used by anyone in real life, i'm only talking about testing purposes.

Reply:
This has a different usecase than the config. Eg: you may want to load the repo you have as a `MemoryRepo`.
The config does not solve the problem, neither it is useful. There may be reason for us to support `ODB_ALTERNATE_BACKENDS`, but that's a separate topic.

> not used right now anywhere and is rather hacky

It's not public. Depends on what you mean by hacky, it's experimental, it's a private API, but depends on stable API/behavior (`Repo()` should be stable, and setting `self.odb.repo` is not hacky at all). There should be room in DVC for experimentation. I agree that it could be improved, but that could be done later too.
If at some point, if `Repo` can replace this, we can get rid of it.

Again, I have to mention that we are stretching the Repo with support for non-local fs, uninitialized repos, etc and I'd prefer to see smaller subclassed classes with a single concern than packing everything in one `Repo` class.

Reply:
> Eg: you may want to load the repo you have as a MemoryRepo.

Right, and you could do that by opening an existing repo and passing a `config=` with a memory cache. I totally agree that MemoryRepo is useful in theory, I just don't like the way it is implemented right now (hacking odb of an existing repo instance) and I don't like it being a dead code that is not currently used or needed anywhere. And it not being public doesn't really mitigate it, as it is still dead code. I don't see the need to merge this right now.

> Again, I have to mention that we are stretching the Repo with support for non-local fs, uninitialized repos, etc and I'd prefer to see smaller subclassed classes with a single concern than packing everything in one Repo class

I disagree, all of those are useful and are currently actively used. Sure, the `__init__` could be cleaner, but still. I don't thing that this justifies adding dead code.